Linear: support sparse_csr/sparse_bsr weights

amjames opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #84897

No new features have been introduced from the CuSparse API, we have the
ability to use more combinations through transposition of inputs
generating a layout swap.

amjames committed:
Improve sparse mm, sparse addmm layout coverage.

This is a first pass at expanding sparse coverage for matrix multiply.
No new features have been introduced from the CuSparse API, we have the
ability to use more combinations through transposition of inputs
generating a layout swap.

[ghstack-poisoned]

cpuhrsch reviewed (changes requested):
[CI failure](https://github.com/pytorch/pytorch/actions/runs/3041310636/jobs/4898470294) looks real

amjames commented:
@cpuhrsch confirmed it is real. I am working the issue. Seems that `torch.zero_` is being called somewhere which asserts csr layout only.

amjames committed:
Update on "Improve sparse mm, sparse addmm layout coverage."


This is a first pass at expanding sparse coverage for matrix multiply.
No new features have been introduced from the CuSparse API, we have the
ability to use more combinations through transposition of inputs
generating a layout swap.

This change enables the pathways required for `torch.nn.functional.linear` with `sparse_csr` and `sparse_bsr` weights. 

[ghstack-poisoned]

amjames commented on torch/testing/_internal/opinfo/core.py:
```diff
@@ -238,6 +238,9 @@ def _tt(t):
         def to_numpy(t):
             if isinstance(t, torch.Tensor):
+                # handle sparse inputs by converting to dense
+                if not t.layout == torch.strided:
```
This would read better as `t.layout != torch.strided`

pearu commented on aten/src/ATen/native/sparse/cuda/SparseBlasImpl.cpp:
```diff
@@ -835,9 +835,17 @@ void addmm_out_sparse_csr(
   if (mat1.layout() == kSparseBsr && mat2.layout() == kStrided && result.layout() == kStrided) {
     return block_sparse_mm(mat1, mat2, beta, alpha, result);
   }
+  if (mat1.layout() == kStrided && mat2.layout() == kSparseBsc &&
+      result.layout() == kStrided) {
+    return block_sparse_mm(mat2.t(), mat1.t(), beta, alpha, result.t());
+  }
   if (mat1.is_sparse_csr() && mat2.layout() == kStrided && result.layout() == kStrided) {
     return spmm(mat1, mat2, beta, alpha, result);
   }
+  if (mat1.layout() == kStrided && mat2.layout() == kSparseCsc &&
+      result.layout() == kStrided) {
+    return spmm(mat2.t(), mat1.t(), beta, alpha, result.t());
+  }
```
Nit: To facilitate batch inputs, the transpose operations ought to use `.transpose(-2, -1)` instead of `.t()`.

(Furthermore, keeping in mind hybrid sparse support, the transpose method call ought to read `.transpose(nbatchdims, nbatchdims+1)` but perhaps not in this PR as underlying sparse tools may not support hybrid sparse tensors out from the box).

nikitaved replied:
I wonder whether `mT` works as well. I am also not sure whether `mH` works, and whether `transpose + conj` is a view on these tensors. I do not remember seeing these things tested, however...

pearu replied:
yes, `mT` works for all layouts. `mH` works for strided inputs only and is a view.

amjames replied:
While we do not have a ton of support for hybrid sparse compressed it is starting to pop up. We don't have a mm kernel that could handle them though so we should TORCH_CHECK to exclude that possibility.

pearu commented on aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp:
```diff
@@ -597,6 +598,18 @@ Tensor _sparse_csr_mm(const Tensor& mat1, const Tensor& mat2) {
+    //  perform (mat2^T x mat1^T)^T
+    return at::addmm(
+               at::zeros({mat2.size(1), mat1.size(0)}, mat1.options()),
+               mat2.t(),
+               mat1.t(),
+               0.0,
+               1.0)
+        .t();
```
See the comment above about replacing `t()` with `transpose(-2, -1)` for batch inputs.

pearu commented on aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp:
```diff
@@ -579,7 +579,8 @@ Tensor _sparse_csr_mm(const Tensor& mat1, const Tensor& mat2) {
     return _sparse_csr_mm(mat1.to_sparse_csr(), mat2);
   }
-  if (mat1.is_sparse_csr() && mat2.layout() == c10::kStrided) {
+  if ((mat1.is_sparse_csr() || (mat1.layout() == kSparseBsr)) &&
```
```suggestion
  if ((mat1.layout() == kSparseCsr) || (mat1.layout() == kSparseBsr)) &&
```

pearu commented on aten/src/ATen/native/mkl/SparseBlasImpl.cpp:
```diff
@@ -367,6 +367,15 @@ void addmm_out_sparse_csr(
+  if (mat1.layout() == kStrided &&
+      (mat2.layout() == kSparseCsc || mat2.layout() == kSparseBsc)) {
+    return addmm_dense_result(
+        mat2.transpose(0, 1),
+        mat1.transpose(0, 1),
+        beta,
+        alpha,
+        result.transpose(0, 1));
```
Nit:
```suggestion
        mat2.transpose(-2, -1),
        mat1.transpose(-2, -1),
        beta,
        alpha,
        result.transpose(-2, -1));
```

nikitaved replied:
The code of the method looks like it might benefit from some refactoring...

pearu reviewed (changes requested):
Overall, looks good! I have a number of nits and suggestions.

pearu commented:
The CI failure is due to https://github.com/pytorch/pytorch/issues/84998 and likely triggered by `_sparse_mm_csr_out` code:
```
zero = at::zeros({mat1.size(0), mat2.size(1)}, mat2.options().layout(result.layout()));
```

amjames commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -3453,6 +3453,9 @@ def sample_inputs_linear(self, device, dtype, requires_grad, **kwargs):
     create_tensor = partial(make_tensor, device=device, dtype=dtype,
                             requires_grad=requires_grad, low=-2, high=2)
 
+    generate_csr_samples = not (dtype in integral_types() or requires_grad)
```
We do not have support for backward yet, that will be a follow up.

nikitaved replied:
Maybe `dtype not in integral_types() and not requires_grad` is easier to parse?

amjames commented on torch/testing/_internal/opinfo/core.py:
```diff
@@ -252,7 +255,8 @@ def to_numpy(t):
     def noncontiguous(self):
         def to_noncontiguous(t):
-            if isinstance(t, torch.Tensor):
+            # currently sparse layouts have continuity invariants that make this impossible
```
It would be preferable to skip or x fail in this situation, unfortunately I don't know of any mechanism for conditionally skipping or xfailing based on attributes of the test sample.

Right now we will silently skip generation of non-contiguous sparse inputs but we still exercise the op with non-contiguous dense inputs.

The best intermediate I can think of would be to put code here which generates contiguous sparse and validates them. Expecting validation to fail, should validation of that transformed sparse start passing at some point an exception is raised indicating it needs to be updated. 

That would be a lot of work, as the validation checks are non-trivial.

nikitaved commented on aten/src/ATen/native/mkl/SparseBlasImpl.cpp:
```diff
@@ -367,6 +367,15 @@ void addmm_out_sparse_csr(
         alpha,
```
Judging from the lines above `addmm_dense_result` should also work with `kSparseBsr`.

nikitaved replied:
I wonder, is it that hard to add these transposition flags?

amjames replied:
Probably best left to a follow up.

nikitaved commented on aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp:
```diff
@@ -550,7 +549,8 @@ Tensor& _sparse_csr_mm_out(
     zero = at::empty({mat1.size(0), mat2.size(1)}, mat2.options());
   } else {
-    zero = at::zeros({mat1.size(0), mat2.size(1)}, mat2.options());
+    zero = at::zeros(
+        {mat1.size(0), mat2.size(1)}, mat2.options().layout(result.layout()));
```
We can just create an "empty" tensor to not rely on these `zeros`, no?

pearu replied:
yes: https://github.com/pytorch/pytorch/pull/85030

nikitaved commented on aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp:
```diff
@@ -597,6 +598,18 @@ Tensor _sparse_csr_mm(const Tensor& mat1, const Tensor& mat2) {
         0.0,
         1.0);
   }
+  if (mat1.layout() == c10::kStrided &&
+      (mat2.layout() == kSparseCsc || mat2.layout() == kSparseBsc)) {
+    // csc.transpose() -> csr or bsc.transpose() -> bsr
```
Same as the comment above. `addmm` should work with `mat2.layout == kSparseBsr` which follows from your added lines.

pearu commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -3464,6 +3467,35 @@ def sample_inputs_linear(self, device, dtype, requires_grad, **kwargs):
 
         bias = create_tensor([out_feat])
         sample_inputs.append(SampleInput(input_tensor, args=(weight, bias)))
+
+    if generate_csr_samples:
+        for has_bias, (in_feat, out_feat), batch_shape in \
+                itertools.product([True, False], features_options, batch_options):
+            input_tensor = create_tensor(batch_shape + [in_feat])
+            weight = create_tensor([out_feat, in_feat]).to_sparse_csr()
+            if not has_bias:
+                sample_inputs.append(SampleInput(input_tensor, args=(weight,)))
+                continue
+
+            bias = create_tensor([out_feat])
+            sample_inputs.append(SampleInput(input_tensor, args=(weight, bias)))
+
+    if generate_bsr_samples:
+        for has_bias, (in_feat, out_feat), batch_shape in \
+                itertools.product([True, False], features_options, batch_options):
+            # for bsr multiply sizes by factor to ensure block-shapes are valid
+            out_feat = out_feat * 16
+            in_feat = in_feat * 16
+            input_tensor = create_tensor(batch_shape + [in_feat])
+            for block_size in (2, 4, 8):
+                weight = create_tensor([out_feat, in_feat]).to_sparse_bsr(block_size, block_size)
+                if not has_bias:
+                    sample_inputs.append(SampleInput(input_tensor, args=(weight,)))
+                    continue
+
+                bias = create_tensor([out_feat])
+                sample_inputs.append(SampleInput(input_tensor, args=(weight, bias)))
+
```
I believe that my implementation suggestion is identical to yours: `weights` are either strided or sparse_csr/bsr. The main difference is that three loops are joined into one.

amjames replied:
Sorry, I missed that detail.

amjames committed:
Update on "Linear: support sparse_csr/sparse_bsr weights"


No new features have been introduced from the CuSparse API, we have the
ability to use more combinations through transposition of inputs
generating a layout swap.


[ghstack-poisoned]

amjames commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -11249,6 +11291,37 @@ def reference_flatten(input, start_dim=0, end_dim=-1):
                DecorateInfo(toleranceOverride({torch.float16: tol(atol=1e-02, rtol=1e-02)}),
                             'TestCudaFuserOpInfo', 'test_nvfuser_correctness'),
            )),
+    OpInfo('nn.functional.linear',
```
After various offline discussion I think this is the best solution for testing ATM. 

The new OpInfo variant generates sparse samples exclusively and adds xfails with are problematic due to the test not yet handling sparsity naturally.

amjames commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -11249,6 +11291,37 @@ def reference_flatten(input, start_dim=0, end_dim=-1):
+               # RuntimeError: Expected all tensors to be on the same device...
+               DecorateInfo(unittest.expectedFailure, "TestCommon", "test_multiple_devices"),
+               # Unsupported Memory format option, Preserve, float32 only?
+               DecorateInfo(unittest.expectedFailure, "TestJit", "test_variant_consistency_jit", dtypes=(torch.float32, ))
```
I am unsure if this failure is legit or not. The test passes for other dtypes.

amjames committed:
Update on "Linear: support sparse_csr/sparse_bsr weights"


No new features have been introduced from the CuSparse API, we have the
ability to use more combinations through transposition of inputs
generating a layout swap.


[ghstack-poisoned]

pearu commented on aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp:
```diff
@@ -457,11 +457,10 @@ Tensor& addmm_out_sparse_compressed_cpu(
       mat1.size(0), "x", mat1.size(1), " and ", mat2.sizes()[0], "x", mat2.sizes()[1], ")");
 
-  if (mat1.layout() == kSparseCsc || mat2.layout() == kSparseCsc) {
+  if (mat1.layout() == kSparseCsc && mat2.layout() == kSparseCsc) {
```
Does this change restrict the existing support matrix of `addmm`?

amjames replied:
No, this condition lead to some valid combinations getting blocked one layer deeper in  `at::native:sparse::impl::mkl::addmm_out_sparse_csr` where more layout handling is done.  

This check is poorly placed and it is too greedy. It forces unnecessary conversions and can cause some strange "layout combination not supported errors".  Suppose you pass `mat1(strided)` `mat2(csc)`, and `self/result (strided)`. The original check would catch causing a `strided->csr` conversion on `mat1`, and an unnecessary and expensive `csc->csr` conversion on `mat2`. When this eventually enters the `impl::mkl` layer you get an error that `strided + csr @ csr` is not supported. Instead now the `strided + strided @ CSC` combination is properly handled by evaluating the transposed operation which is `strided + CSR @ strided` which we have support for. 

In fact, coupled with the other changes in the `mkl` interface in this pr relaxing the condition here expands support for matrix addmm.

pearu replied:
Makes sense.

pearu reviewed (changes requested):
I have various nits and typos, otherwise looks good!
Thanks, @amjames!

amjames commented on aten/src/ATen/native/sparse/cuda/SparseBlasImpl.cpp:
```diff
@@ -835,12 +835,30 @@ void addmm_out_sparse_csr(
   if (mat1.layout() == kSparseBsr && mat2.layout() == kStrided && result.layout() == kStrided) {
```
I will try to clean this up a bit

amjames commented:
I have split this up, closing in favor of new stack.
